PM related changes - V2

ranj063 opened the pull request:
1. snd_pcm_suspend() is going to be a static function. So move to using snd_pcm_suspend_all() instead.  But to avoid any race conditions in situations that the streams are stopped prior to suspend (as in the case of chrome), only suspend those pcm's which have at least one running stream.
2. send ipc to stop stream when a suspend trigger is received.
3. Because snd_pcm_suspend_all() suspends all streams in a pcm, unset restore_stream in hw_params to avoid duplicating the call to hw_params for streams that were suspended even though they werent running.

CHanges in V2:

1. add pm ops for bxT_pcm521x machine driver
2. remove suspend streams in SOF suspend callback

ranj063 commented:
fixes https://github.com/thesofproject/linux/issues/570

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -198,29 +198,30 @@ static void sof_suspend_streams(struct snd_sof_dev *sdev)
+	int dir, state, playback_dir, capture_dir;
 
-	/* suspend all running streams */
+	/* suspend all running/paused streams */
```
@ranj063 if you look at recent patches from Takashi Iwai on alsa-devel, tracked with issue https://github.com/thesofproject/linux/issues/550, the direction is to make snd_pcm_suspend a static function and let the framework deal with PCM suspend. Is this whole sequence really necessary?

ranj063 replied:
@plbossart I'll have to check how the framework handles pcm suspend. let me look into it today.

ranj063 replied:
@plbossart I've updated the PR now. And we should be good when the changes in Takashi's proposal are merged.

ranj063 commented:
> I am really confused by the approach here.
> 
> ASoC defines a set of routines to deal with PM. Most drivers seem to use them, and _some_ Intel boards do
> 
> ```
> intel/boards/bxt_da7219_max98357a.c:            .pm = &snd_soc_pm_ops,
> intel/boards/bxt_rt298.c:               .pm = &snd_soc_pm_ops,
> intel/boards/byt-max98090.c:            .pm = &snd_soc_pm_ops,
> intel/boards/byt-rt5640.c:              .pm = &snd_soc_pm_ops,
> intel/boards/glk_rt5682_max98357a.c:            .pm = &snd_soc_pm_ops,
> intel/boards/kbl_da7219_max98357a.c:            .pm = &snd_soc_pm_ops,
> intel/boards/kbl_da7219_max98927.c:             .pm = &snd_soc_pm_ops,
> intel/boards/kbl_rt5660.c:              .pm = &snd_soc_pm_ops,
> intel/boards/kbl_rt5663_max98927.c:             .pm = &snd_soc_pm_ops,
> intel/boards/kbl_rt5663_rt5514_max98927.c:              .pm = &snd_soc_pm_ops,
> intel/boards/skl_hda_dsp_generic.c:             .pm = &snd_soc_pm_ops,
> intel/boards/skl_nau88l25_max98357a.c:          .pm = &snd_soc_pm_ops,
> intel/boards/skl_nau88l25_ssm4567.c:            .pm = &snd_soc_pm_ops,
> intel/boards/skl_rt286.c:               .pm = &snd_soc_pm_ops,
> intel/boards/sof_rt5682.c:              .pm = &snd_soc_pm_ops,
> ```
> So my problem is why we need to deal with stream suspend/resume if we already have a framework support for it?

@plbossart I might be totally wrong here but snd_soc_pm_ops looks like it was defined for the ASoC platform driver. And I believe the machine driver is using the same pm ops instead of defining a new one. 

To address you question of handling stream suspend, if I don't do it in SOF, nothing ever happens. I checked the skylake driver and I dont see that it handles stream suspend at all. I am not sure how else to get it working.

plbossart commented:
>     So my problem is why we need to deal with stream suspend/resume if
>     we already have a framework support for it?
>
> @plbossart <https://github.com/plbossart> I might be totally wrong 
> here but snd_soc_pm_ops looks like it was defined for the ASoC 
> platform driver. And I believe the machine driver is using the same pm 
> ops instead of defining a new one.
>
> To address you question of handling stream suspend, if I don't do it 
> in SOF, nothing ever happens. I checked the skylake driver and I dont 
> see that it handles stream suspend at all. I am not sure how else to 
> get it working.
>
git grep snd_soc_pm_ops tells me it's a card/machine driver level. And 
maybe we are missing it e.g. in all our SOF machine drivers...You will 
note it's enabled in all Chromebooks machine drivers so that's a hint.

It was historically not enabled for Atom/haswell/broadwell devices since 
there were weird things to deal with, but in our case I really wonder if 
all we need to do is just save/restore context but leave streams alone.

That said doing most of the work at the card level is interesting, and 
maybe not quite aligned with the parent/child relationship we want to have.

ranj063 commented:
> So my problem is why we need to deal with stream suspend/resume if we already have a framework support for it? @plbossart <https://github.com/plbossart> I might be totally wrong here but snd_soc_pm_ops looks like it was defined for the ASoC platform driver. And I believe the machine driver is using the same pm ops instead of defining a new one. To address you question of handling stream suspend, if I don't do it in SOF, nothing ever happens. I checked the skylake driver and I dont see that it handles stream suspend at all. I am not sure how else to get it working.
> git grep snd_soc_pm_ops tells me it's a card/machine driver level. And maybe we are missing it e.g. in all our SOF machine drivers...You will note it's enabled in all Chromebooks machine drivers so that's a hint. It was historically not enabled for Atom/haswell/broadwell devices since there were weird things to deal with, but in our case I really wonder if all we need to do is just save/restore context but leave streams alone. That said doing most of the work at the card level is interesting, and maybe not quite aligned with the parent/child relationship we want to have.

@plbossart the haswell driver is the only one that deals with stream suspend. I'll need to unwrap the mysetery behind how the skylake driver suspends streams without the call to snd_pcm_suspend_all(). Let me talk to some folks tonight and get some information.

Also, on the chromebook, suspending from the UI works just fine. But it fails from the command line. So even though the machine driver includes the pm ops, it doesnt seem to help with the command line case.

ranj063 commented:
@plbossart i finally got the suspend to work with audio playback from cmd line by making this change and adding snd_soc_pm_ops to the bxt_pcm_512x machine driver:
````diff
diff --git a/sound/soc/soc-core.c b/sound/soc/soc-core.c
index fc7dadfbc10f..93c188fb1400 100644
--- a/sound/soc/soc-core.c
+++ b/sound/soc/soc-core.c
@@ -16,6 +16,7 @@
 //   o More testing with other codecs/machines.
 //   o Add more codecs and platforms to ensure good API coverage.
 //   o Support TDM on PCM and I2S
+#define DEBUG
 
 #include <linux/module.h>
 #include <linux/moduleparam.h>
@@ -471,12 +472,13 @@ int snd_soc_suspend(struct device *dev)
                                drv->ops->digital_mute(dai, 1);
                }
        }
+       dev_dbg(card->dev, "ranjani: suspending all pcm's %d\n", card->num_rtd);
 
        /* suspend all pcms */
        for_each_card_rtds(card, rtd) {
-               if (rtd->dai_link->ignore_suspend)
+               if (rtd->dai_link->ignore_suspend || rtd->dai_link->no_pcm)
                        continue;
-
+               dev_dbg(card->dev, "ranjani: suspending all pcm's\n");
                snd_pcm_suspend_all(rtd->pcm);
        }
````
Basically, the system was rebooting because the rtd_list contained BE dai links and sending the snd_pcm_suspend_all(rtd->pcm) for those BE links fails miserably.

So now the question is:
1. Should the rtd_list contain BE dai_lnks?
2. If so, how do we handle this case? I thought the trigger action for BE dai links does nothing. But looks like there's more to it than meets the eye.
3. If not, why are they getting added in SOF?

@keyonjie could you also comment?

keyonjie commented:
> @plbossart i finally got the suspend to work with audio playback from cmd line by making this change and adding snd_soc_pm_ops to the bxt_pcm_512x machine driver:
> 
> diff --git a/sound/soc/soc-core.c b/sound/soc/soc-core.c
> index fc7dadfbc10f..93c188fb1400 100644
> --- a/sound/soc/soc-core.c
> +++ b/sound/soc/soc-core.c
> @@ -16,6 +16,7 @@
> // o More testing with other codecs/machines.
> // o Add more codecs and platforms to ensure good API coverage.
> // o Support TDM on PCM and I2S
> +#define DEBUG
> 
> #include <linux/module.h>
> #include <linux/moduleparam.h>
> @@ -471,12 +472,13 @@ int snd_soc_suspend(struct device *dev)
> drv->ops->digital_mute(dai, 1);
> }
> }
> 
> * ```
>     dev_dbg(card->dev, "ranjani: suspending all pcm's %d\n", card->num_rtd);
>   
>     /* suspend all pcms */
>     for_each_card_rtds(card, rtd) {
>   ```
> 
> * ```
>             if (rtd->dai_link->ignore_suspend)
>   ```
> 
> * ```
>             if (rtd->dai_link->ignore_suspend || rtd->dai_link->no_pcm)
>                     continue;
>   ```
> 
> 
> * ```
>             dev_dbg(card->dev, "ranjani: suspending all pcm's\n");
>             snd_pcm_suspend_all(rtd->pcm);
>     }
>   ```
> 
> Basically, the system was rebooting because the rtd_list contained BE dai links and sending the snd_pcm_suspend_all(rtd->pcm) for those BE links fails miserably.
> 
> So now the question is:
> 
> 1. Should the rtd_list contain BE dai_lnks?
> 2. If so, how do we handle this case? I thought the trigger action for BE dai links does nothing. But looks like there's more to it than meets the eye.
> 3. If not, why are they getting added in SOF?
> 
> @keyonjie could you also comment?

@ranj063 have you printed out rtd->pcm here for BE dai_links, I just went though ASoC code quickly, we are creating internal PCM for BE dai_links, so suspend them looks no problem to me, these internal PCMs don't have trigger ops so trigger suspend should do nothing there.

ranj063 commented:
@keyonjie its failing precisely because the internal pcm substream's dont have any ops set. In snd_pcm_do_suspend() the call to substream->ops->trigger is invalid because there are no ops set and this causes a null pointer exception that causes the system to reboot.

plbossart commented:
Nice catch. I don't think this code ever worked :-(

````c
	if (rtd->dai_link->no_pcm) {
		if (playback)
			pcm->streams[SNDRV_PCM_STREAM_PLAYBACK].substream->private_data = rtd;
		if (capture)
			pcm->streams[SNDRV_PCM_STREAM_CAPTURE].substream->private_data = rtd;
		goto out; //<<<< this bypasses all the ops initializations!!
	}

	/* ASoC PCM operations */
	if (rtd->dai_link->dynamic) {
		rtd->ops.open		= dpcm_fe_dai_open;
		rtd->ops.hw_params	= dpcm_fe_dai_hw_params;
		rtd->ops.prepare	= dpcm_fe_dai_prepare;
		rtd->ops.trigger	= dpcm_fe_dai_trigger;
		rtd->ops.hw_free	= dpcm_fe_dai_hw_free;
		rtd->ops.close		= dpcm_fe_dai_close;
		rtd->ops.pointer	= soc_pcm_pointer;
		rtd->ops.ioctl		= soc_pcm_ioctl;
	} else {
		rtd->ops.open		= soc_pcm_open;
		rtd->ops.hw_params	= soc_pcm_hw_params;
		rtd->ops.prepare	= soc_pcm_prepare;
		rtd->ops.trigger	= soc_pcm_trigger;
		rtd->ops.hw_free	= soc_pcm_hw_free;
		rtd->ops.close		= soc_pcm_close;
		rtd->ops.pointer	= soc_pcm_pointer;
		rtd->ops.ioctl		= soc_pcm_ioctl;
	}

	for_each_rtdcom(rtd, rtdcom) {
		const struct snd_pcm_ops *ops = rtdcom->component->driver->ops;

		if (!ops)
			continue;

		if (ops->ack)
			rtd->ops.ack		= soc_rtdcom_ack;
		if (ops->copy_user)
			rtd->ops.copy_user	= soc_rtdcom_copy_user;
		if (ops->copy_kernel)
			rtd->ops.copy_kernel	= soc_rtdcom_copy_kernel;
		if (ops->fill_silence)
			rtd->ops.fill_silence	= soc_rtdcom_fill_silence;
		if (ops->page)
			rtd->ops.page		= soc_rtdcom_page;
		if (ops->mmap)
			rtd->ops.mmap		= soc_rtdcom_mmap;
	}

	if (playback)
		snd_pcm_set_ops(pcm, SNDRV_PCM_STREAM_PLAYBACK, &rtd->ops);

	if (capture)
		snd_pcm_set_ops(pcm, SNDRV_PCM_STREAM_CAPTURE, &rtd->ops);

	for_each_rtdcom(rtd, rtdcom) {
		component = rtdcom->component;

		if (!component->driver->pcm_new)
			continue;

		ret = component->driver->pcm_new(rtd);
		if (ret < 0) {
			dev_err(component->dev,
				"ASoC: pcm constructor failed: %d\n",
				ret);
			return ret;
		}
	}

	pcm->private_free = soc_pcm_private_free;
out:
	dev_info(rtd->card->dev, "%s <-> %s mapping ok\n",
		 (rtd->num_codecs > 1) ? "multicodec" : rtd->codec_dai->name,
		 cpu_dai->name);
	return ret;
````

ranj063 commented:
@plbossart @keyonjie checking for ops !- NULL before calling ops->trigger seems to work for getting suspend to work in the case of suspending during I2S playback. But there is still some more work to be done for the case of suspending during HDMI playback. I can see that the skylake driver saves the dpibr and lpib which we dont seem to do. I'm not an expert here. So I could use your help Keyon. I'll post my patches soon.

plbossart commented:
> But there is still some more work to be done for the case of suspending during HDMI playback. I can see that the skylake driver saves the dpibr and lpib which we dont seem to do. I'm not an expert here. So I could use your help Keyon. I'll post my patches soon.

There should be no difference between outputs? I guess saving/restoring DPIB/LPIB is tied to the effort to restart playback exactly from where the stream stopped, supposedly the DMA can deal with context. I don't think this ever worked and I don't see who would care if you restart 10 days later from the same sample you suspended on. This makes no sense from an audio perspective given the mute ramping that's needed anyways to avoid pops/clicks.

ranj063 commented:
@plbossart I've updated this PR now. I've added the pm ops to the bxt_pcm512x machine driver. 

I think this should fix the following issues:
https://github.com/thesofproject/linux/issues/592
https://github.com/thesofproject/linux/issues/489 - fixed partially as only suspending during I2S playback is fixed. suspend during HDMI playback is still broken.

ranj063 commented:
> > But there is still some more work to be done for the case of suspending during HDMI playback. I can see that the skylake driver saves the dpibr and lpib which we dont seem to do. I'm not an expert here. So I could use your help Keyon. I'll post my patches soon.
> 
> There should be no difference between outputs? I guess saving/restoring DPIB/LPIB is tied to the effort to restart playback exactly from where the stream stopped, supposedly the DMA can deal with context. I don't think this ever worked and I don't see who would care if you restart 10 days later from the same sample you suspended on. This makes no sense from an audio perspective given the mute ramping that's needed anyways to avoid pops/clicks.

@plbossart but it seems to working with the skylake driver. So my guess is we are probably just missing some bits.

plbossart commented:
On 1/24/19 12:52 PM, Ranjani Sridharan wrote:
>
>         But there is still some more work to be done for the case of
>         suspending during HDMI playback. I can see that the skylake
>         driver saves the dpibr and lpib which we dont seem to do. I'm
>         not an expert here. So I could use your help Keyon. I'll post
>         my patches soon.
>
>     There should be no difference between outputs? I guess
>     saving/restoring DPIB/LPIB is tied to the effort to restart
>     playback exactly from where the stream stopped, supposedly the DMA
>     can deal with context. I don't think this ever worked and I don't
>     see who would care if you restart 10 days later from the same
>     sample you suspended on. This makes no sense from an audio
>     perspective given the mute ramping that's needed anyways to avoid
>     pops/clicks.
>
> @plbossart <https://github.com/plbossart> but it seems to working with 
> the skylake driver. So my guess is we are probably just missing some bits.
>
What works? and what does 'work' mean?

The Skylake machine drivers all seem to rely on soc_pm_ops so shouldn't 
have anything specific to suspend streams?

ranj063 commented:
> On 1/24/19 12:52 PM, Ranjani Sridharan wrote: But there is still some more work to be done for the case of suspending during HDMI playback. I can see that the skylake driver saves the dpibr and lpib which we dont seem to do. I'm not an expert here. So I could use your help Keyon. I'll post my patches soon. There should be no difference between outputs? I guess saving/restoring DPIB/LPIB is tied to the effort to restart playback exactly from where the stream stopped, supposedly the DMA can deal with context. I don't think this ever worked and I don't see who would care if you restart 10 days later from the same sample you suspended on. This makes no sense from an audio perspective given the mute ramping that's needed anyways to avoid pops/clicks. @plbossart <https://github.com/plbossart> but it seems to working with the skylake driver. So my guess is we are probably just missing some bits.
> What works? and what does 'work' mean? The Skylake machine drivers all seem to rely on soc_pm_ops so shouldn't have anything specific to suspend streams?

@plbossart I mean suspending the system during HDMI playback works with the skylake driver. But in our case, we run into IPC errors. I will post the logs on the issue soon.

plbossart commented:
>     What works? and what does 'work' mean? The Skylake machine drivers
>     all seem to rely on soc_pm_ops so shouldn't have anything specific
>     to suspend streams?
>
> @plbossart <https://github.com/plbossart> I mean suspending the system 
> during HDMI playback works with the skylake driver. But in our case, 
> we run into IPC errors. I will post the logs on the issue soon.
>
That's a different story, I get IPC errors even without HDMI on Up2 :-(. 
Something is seriously broken in our IPC handling, it's just completely 
unpredictable when changing configurations and/or timing on the host side.
>

bardliao commented:
@ranj063 I checked the code flow today. substream->ops will be assigned in soc_new_pcm().
But indeed substream->ops is null when I try to suspend during i2s playback.
So, I am wondering how could it happen.
https://elixir.bootlin.com/linux/latest/source/sound/core/pcm_lib.c#L485
```
void snd_pcm_set_ops(struct snd_pcm *pcm, int direction,
		     const struct snd_pcm_ops *ops)
{
	struct snd_pcm_str *stream = &pcm->streams[direction];
	struct snd_pcm_substream *substream;
	
	for (substream = stream->substream; substream != NULL; substream = substream->next)
		substream->ops = ops;
}
```
https://elixir.bootlin.com/linux/latest/source/sound/soc/soc-pcm.c#L3136
```
	if (playback)
		snd_pcm_set_ops(pcm, SNDRV_PCM_STREAM_PLAYBACK, &rtd->ops);

	if (capture)
		snd_pcm_set_ops(pcm, SNDRV_PCM_STREAM_CAPTURE, &rtd->ops);
```

bardliao commented:
@ranj063 Sorry I was wrong. It will goto out if (rtd->dai_link->no_pcm).
https://elixir.bootlin.com/linux/latest/source/sound/soc/soc-pcm.c#L3087
That's way substream->ops is not set.
So an alternative way for "ALSA: PCM: check if ops are defined before using it" is
```
--- a/sound/soc/soc-core.c
+++ b/sound/soc/soc-core.c
@@ -477,6 +477,8 @@ int snd_soc_suspend(struct device *dev)
                if (rtd->dai_link->ignore_suspend)
                        continue;

+             if (rtd->dai_link->no_pcm)
+                     continue;
                snd_pcm_suspend_all(rtd->pcm);
        }
```
But I am wondering how did it work before?

plbossart commented:
> @ranj063 Sorry I was wrong. It will goto out if (rtd->dai_link->no_pcm).
> https://elixir.bootlin.com/linux/latest/source/sound/soc/soc-pcm.c#L3087
> That's way substream->ops is not set.

Yes, this is aligned with my analysis (posted above)

> So an alternative way for "ALSA: PCM: check if ops are defined before using it" is
> 
> ```
> --- a/sound/soc/soc-core.c
> +++ b/sound/soc/soc-core.c
> @@ -477,6 +477,8 @@ int snd_soc_suspend(struct device *dev)
>                 if (rtd->dai_link->ignore_suspend)
>                         continue;
> 
> +             if (rtd->dai_link->no_pcm)
> +                     continue;
>                 snd_pcm_suspend_all(rtd->pcm);
>         }
> ```

Agree, I think it's a better fix since it's contained within ASoC. 

> But I am wondering how did it work before?

It never did...

ranj063 commented:
> > @ranj063 Sorry I was wrong. It will goto out if (rtd->dai_link->no_pcm).
> > https://elixir.bootlin.com/linux/latest/source/sound/soc/soc-pcm.c#L3087
> > That's way substream->ops is not set.
> 
> Yes, this is aligned with my analysis (posted above)
> 
> > So an alternative way for "ALSA: PCM: check if ops are defined before using it" is
> > ```
> > --- a/sound/soc/soc-core.c
> > +++ b/sound/soc/soc-core.c
> > @@ -477,6 +477,8 @@ int snd_soc_suspend(struct device *dev)
> >                 if (rtd->dai_link->ignore_suspend)
> >                         continue;
> > 
> > +             if (rtd->dai_link->no_pcm)
> > +                     continue;
> >                 snd_pcm_suspend_all(rtd->pcm);
> >         }
> > ```
> 
> Agree, I think it's a better fix since it's contained within ASoC.
> 
> > But I am wondering how did it work before?
> 
> It never did...
@bardliao @plbossart  
if (rtd->dai_link_no_pcm)
    continue;

is how I found the problem but I do not think this will be accepted as a proper solution. My guess is codec<->codec links might need this. 

The right fix is to check for the missing ops. We cant call ops without it being valid. 

As for the question about "how did it work before", I spent my whole last night debugging whats going on in skylake. The substream->status->state for the BE dai link pcm in the case of skylake is already in SUSPENDED state. So snd_pcm_suspend() returns with -EBUSY error which of course is ignored.I am not sure why is it SUSPENDED in the first place, if the audio is playing it should be RUNNING and thats what state the pcm is in SOF.

ranj063 commented:
@plbossart @bardliao @keyonjie after a bit of digging around, I figured out how this is working for skylake. 
Basically, because the FE dai links and BE dai links are both defined in the machine driver, the FE dai_links get added to the rtd_list first. So when we call snd_pcm_suspend_all() , the FE dai links get suspended first. And then when we get to the BE dai links whose runtime is actually set to the FW dai link's pcm when the stream is started, is actually already suspended, so snd_pcm_suspend returns with a -EBUSY. This error is ignored.

But in the case of SOF, the BE dai links are added  to the rtd_list first during machine driver probe and then the FE dai links are added during topology loading. So this results in the BE dai links pcm getting suspended first which leads to null pointer exception for ops->trigger. 

Given all this, I think the best fix is what is in this PR. Appreciate your comments.

plbossart commented:
On 1/25/19 6:45 PM, Ranjani Sridharan wrote:
>
> @plbossart <https://github.com/plbossart> @bardliao 
> <https://github.com/bardliao> @keyonjie <https://github.com/keyonjie> 
> after a bit of digging around, I figured out how this is working for 
> skylake.
> Basically, because the FE dai links and BE dai links are both defined 
> in the machine driver, the FE dai_links get added to the rtd_list 
> first. So when we call snd_pcm_suspend_all() , the FE dai links get 
> suspended first. And then when we get to the BE dai links whose 
> runtime is actually set to the FW dai link's pcm when the stream is 
> started are actually already suspended.
>
I don't think this is fully correct, if you look at skl_hda_dsp_common.c 
the front-ends are set with the topology. the hard-coding of the FEs in 
machine drivers is historical, but that's not mandatory.

It'd be good if you can install the latest kernel on your KBL XPS13 and 
check for yourself.

> But in the case of SOF, the BE dai links are added to the rtd_list 
> first during machine driver probe and then the FE dai links are added 
> during topology loading. So this results in the BE dai links pcm 
> getting suspended first which leads to null pointer exception for 
> ops->trigger.
>
> Given all this, I think the best fix is what is in this PR. Appreciate 
> your comments.
>
I think we need to add the check in ASoC rather than in the core.

ranj063 commented:
> On 1/25/19 6:45 PM, Ranjani Sridharan wrote: @plbossart <https://github.com/plbossart> @bardliao <https://github.com/bardliao> @keyonjie <https://github.com/keyonjie> after a bit of digging around, I figured out how this is working for skylake. Basically, because the FE dai links and BE dai links are both defined in the machine driver, the FE dai_links get added to the rtd_list first. So when we call snd_pcm_suspend_all() , the FE dai links get suspended first. And then when we get to the BE dai links whose runtime is actually set to the FW dai link's pcm when the stream is started are actually already suspended.
> I don't think this is fully correct, if you look at skl_hda_dsp_common.c the front-ends are set with the topology. the hard-coding of the FEs in machine drivers is historical, but that's not mandatory. It'd be good if you can install the latest kernel on your KBL XPS13 and check for yourself.
@plbossart it is not mandatory I agree. I was only pointing out how this is working on the GLK chromebook with the SKL driver. 

Basically, the current flow only works if the FE dai links  are added to the rtd_list first and then followed by the BE dai links. In SOF, the order is the opposite and that is why we hit the null pointer with ops-?trigger() for the BE dai links.


> But in the case of SOF, the BE dai links are added to the rtd_list first during machine driver probe and then the FE dai links are added during topology loading. So this results in the BE dai links pcm getting suspended first which leads to null pointer exception for ops->trigger. Given all this, I think the best fix is what is in this PR. Appreciate your comments.
> I think we need to add the check in ASoC rather than in the core.

We could check in ASoC but I have a feeling it will not be an accepted solution. What about hostless PCM's and codec<->codec links? There will be no way to add a trigger for such cases?

plbossart commented:
On 1/28/19 10:53 AM, Ranjani Sridharan wrote:
>
>     On 1/25/19 6:45 PM, Ranjani Sridharan wrote: @plbossart
>     <https://github.com/plbossart> https://github.com/plbossart
>     @bardliao <https://github.com/bardliao>
>     https://github.com/bardliao @keyonjie
>     <https://github.com/keyonjie> https://github.com/keyonjie after a
>     bit of digging around, I figured out how this is working for
>     skylake. Basically, because the FE dai links and BE dai links are
>     both defined in the machine driver, the FE dai_links get added to
>     the rtd_list first. So when we call snd_pcm_suspend_all() , the FE
>     dai links get suspended first. And then when we get to the BE dai
>     links whose runtime is actually set to the FW dai link's pcm when
>     the stream is started are actually already suspended.
>     I don't think this is fully correct, if you look at
>     skl_hda_dsp_common.c the front-ends are set with the topology. the
>     hard-coding of the FEs in machine drivers is historical, but
>     that's not mandatory. It'd be good if you can install the latest
>     kernel on your KBL XPS13 and check for yourself.
>     @plbossart <https://github.com/plbossart> it is not mandatory I
>     agree. I was only pointing out how this is working on the GLK
>     chromebook with the SKL driver.
>
> Basically, the current flow only works if the FE dai links are added 
> to the rtd_list first and then followed by the BE dai links. In SOF, 
> the order is the opposite and that is why we hit the null pointer with 
> ops-?trigger() for the BE dai links.
>
Sorry, you lost me here. Why would the order of the insertion in 
rtd_list matter? We saw earlier that there was a goto that skips the 
.ops init.

>     But in the case of SOF, the BE dai links are added to the rtd_list
>     first during machine driver probe and then the FE dai links are
>     added during topology loading. So this results in the BE dai links
>     pcm getting suspended first which leads to null pointer exception
>     for ops->trigger. Given all this, I think the best fix is what is
>     in this PR. Appreciate your comments.
>     I think we need to add the check in ASoC rather than in the core.
>
> We could check in ASoC but I have a feeling it will not be an accepted 
> solution. What about hostless PCM's and codec<->codec links? There 
> will be no way to add a trigger for such cases?
>
I see, so your point is that we only need to fix the PCM trigger case 
and avoid having any side effects on other cases that make use of the 
.no_pcm field, right?

ranj063 commented:
@plbossart this one is going to hard to explain but let me try to tell the whole story here: 

First the story of the GLK sst driver:

BE dai links only have internal PCM's whose ops are not set when soc_pcm_new() is called.
When the stream is started, the BE dai link PCM runtime is set to that of the FE dai link PCM runtime so that they are connected. 

Now, if we rewind a little bit to when the rtd_list is updated, in the case of the SST driver, this is done in 2 places: 

1. All dai links defined in the machine driver (it seems to be called pre_links)
2. All dai links defined in topology. 

If you look at the bxt_da7219_max98357a.c, the broxton_dais[] has the FE dai links followed by the BE dai links. So the rtd_list would contain them in the same order. 

Coming back to when we suspend, the snd_soc_suspend() callback in soc-core.c suspends each item in rtd_list in the loop for_each_card_rtds(card, rtd) {}. This means the FE dai links will be suspended first. And when it gets to the BE dai links, since the PCM runtime is the same as that of the FE dai link, it will already be suspended. So snd_pcm_suspend_all() for this BE dai link returns -EBUSY. This is ignored and all is well.

Moving on to the SOF side of the story:

The first part remains the same. BE dai links do no have ops set and their runtime PCM is set to that of the FE dai link when the stream is started. 

But, the rtd_list would be updated in this order:
1. BE dai links from the machine driver ( the pre_links). The FE dai links in the machine driver are ignored
2. Followed by the FE dai links from topology. 

Now, when we suspend, snd_soc_suspend() will suspend the runtime PCM's in the rtd_list() in the order the dai links were added in the loop for_each_card_rtds(card, rtd) {}. So the BE dai links go first and since their state is RUNNING, it will attempt to trigger suspend ops for the BE dai link and throw a null pointer exception. 


Please let me know if this makes sense.

RanderWang commented:
update my status:

I checked FW and found that the setting of link dma is correct but link dma didn't work.
I do two modifications to make HDMI resume back successfully.
(1) make sure trigger function of be dai is called by resume process (hda_link_pcm_trigger)
     Now it is not called because SND_SOC_DPCM_STATE_SUSPEND is never assigned to state
````
dpcm_be_dai_trigger() 
{
             ...
		case SNDRV_PCM_TRIGGER_RESUME:
			if ((be->dpcm[stream].state != SND_SOC_DPCM_STATE_SUSPEND))
				continue;
			ret = dpcm_do_trigger(dpcm, be_substream, cmd);
			if (ret)
				return ret;
            ...
}
````

(2) in hda_link_pcm_trigger restore hw params for BE dai. In our FE pcm (sof_pcm_trigger), we do this.

ranj063 commented:
@RanderWang thanks for your analysis. Do you have a patch you can send ?

RanderWang commented:
> @RanderWang thanks for your analysis. Do you have a patch you can send ?

The modification is not fit for PR,  it is just for experiment.

stevyan commented:
**Summary:**
The bug [linux#592](https://github.com/thesofproject/linux/issues/592) can be fixed with this pull request based on daily build version.

**Test Env:**
sof master: 365b132e
kernel sof-dev: [pr582-8](https://github.com/thesofproject/linux/pull/582)
tplg: sof-apl-nocodec.tplg

**Reference Info:**

> **ubuntu@cd-apl-up2-nocodec-04:\~$** ps aux | grep aplay
> ubuntu 2624 0.0 0.1 76460 5908 pts/0 S 23:58 0:00 aplay --dump-hw-params -Dhw:0,0 -r
> 48000 -c 2 -f S16_LE /dev/zero
> ubuntu 2640 0.0 0.0 21536 1052 pts/0 S+ 23:59 0:00 grep --color=auto aplay
> **ubuntu@cd-apl-up2-nocodec-04:\~$** sudo rtcwake -m mem -s 2
> rtcwake: assuming RTC uses UTC ...
> rtcwake: wakeup from "mem" using /dev/rtc0 at Sun Mar 18 20:21:51 2012
> Suspended. Trying resume. **ubuntu@cd-apl-up2-nocodec-04:\~$** Done.

ranj063 commented:
@plbossart please dont merge this PR as is. I want to make one more change here. I will submit the change soon.

plbossart commented:
@mengdonglin @xiulipan can you please check the CI status, nothing seems to build.

@ranj063 is this PR candidate for a merge? I am really scratching my head on what PR should go in first, nothing is really stable with kernel oops left and right and what I merge will end-up breaking something else.

RanderWang commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -162,18 +162,24 @@ static int hda_link_hw_params(struct snd_pcm_substream *substream,
 		p_params.link_bps = codec_dai->driver->capture.sig_bits;
 
+	/* save pipe params */
+	rtd->private = params;
```
params is a local variable, you could not save it

ranj063 replied:
@RanderWang I am able to retrieve the right values after resuming. So I am guessing there wasnt anything worng with the saving part.

RanderWang replied:
yes,  I don't know whether  hw_params in HDMI codec is restored by ASoC。You could add a log in HDMI codec to check whether hw_params function is called after resuming

ranj063 replied:
@RanderWang I can check that. But this seems to work fine for the legacy driver. So I'm guesing it should be taken care of.

RanderWang replied:
yes, I check the legacy hda driver two days ago and confirmed that we should restore BE dai link.
I found that legacy hda driver do some work in restoring process. By the way, you could talk with HDA maintainer takashi

RanderWang replied:
you could hear sound after restart. The difference is everything would be set by ASoC again. So something may need to be restored after resuming.

ranj063 commented:
@plbossart updated now. I will submit a separate PR for the link DMA resuming part

lgirdwood commented on sound/soc/intel/boards/bxt_pcm512x.c:
```diff
@@ -305,6 +305,7 @@ static int bxt_pcm512x_probe(struct platform_device *pdev)
 static struct platform_driver bxt_pcm521x_driver = {
 	.driver = {
 		.name = "bxt-pcm512x",
+		.pm = &snd_soc_pm_ops,
```
Probably worth checking if this is missing from the other Intel machine drivers.

plbossart commented:
On 2/1/19 10:12 AM, Liam Girdwood wrote:
>
> *@lgirdwood* commented on this pull request.
>
> @ranj063 <https://github.com/ranj063> One thing, I will be adding some 
> new tokens that will tell SOF PM core not to suspend a PCM. i.e. if 
> PCM is keyword detection PCM. I assume nothing here will prevent this 
> flow ?
>
I'd like to understand how our esteemed colleagues managed to deal with 
a PCM that wasn't suspended without tokens.
> ------------------------------------------------------------------------
>
> In sound/soc/intel/boards/bxt_pcm512x.c 
> <https://github.com/thesofproject/linux/pull/582#discussion_r253105138>:
>
> > @@ -305,6 +305,7 @@ static int bxt_pcm512x_probe(struct platform_device *pdev)
>   static struct platform_driver bxt_pcm521x_driver = {
>   	.driver = {
>   		.name = "bxt-pcm512x",
> +		.pm = &snd_soc_pm_ops,
>
> Probably worth checking if this is missing from the other Intel 
> machine drivers.
>
yes, all Atom machine drivers and wm8804 for sure. All others are fine 
already.
>

ranj063 commented:
> On 2/1/19 10:12 AM, Liam Girdwood wrote: ***@***.**** commented on this pull request. @ranj063 <https://github.com/ranj063> One thing, I will be adding some new tokens that will tell SOF PM core not to suspend a PCM. i.e. if PCM is keyword detection PCM. I assume nothing here will prevent this flow ?
> I'd like to understand how our esteemed colleagues managed to deal with a PCM that wasn't suspended without tokens.
@plbossart @lgirdwood the skylake driver uses the ignore_suspend member in the DMIC backend da link definition in the machine driver to prevent suspending those PCM's. Can we use the same too?

> ------------------------------------------------------------------------ In sound/soc/intel/boards/bxt_pcm512x.c <[#582 (comment)](https://github.com/thesofproject/linux/pull/582#discussion_r253105138)>: > @@ -305,6 +305,7 @@ static int bxt_pcm512x_probe(struct platform_device *pdev) static struct platform_driver bxt_pcm521x_driver = { .driver = { .name = "bxt-pcm512x", +	.pm = &snd_soc_pm_ops, Probably worth checking if this is missing from the other Intel machine drivers.
> yes, all Atom machine drivers and wm8804 for sure. All others are fine already.
> […](#)

ranj063 commented:
> @ranj063 One thing, I will be adding some new tokens that will tell SOF PM core not to suspend a PCM. i.e. if PCM is keyword detection PCM. I assume nothing here will prevent this flow ?

@lgirdwood yes, nothing will prevent that. We dont explicitly call any pcm suspend anymore in SOF. the snd_soc_suspend() callback ignores  BE dai links that have their ignore_suspend member set to true. So thats what we should use I think.
